cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,543 @@
+    /// equivalence class, so we can find all the binding sites in that class given any one of them.
+    pub equals: HashMap<BindingId, BindingId>,
+    /// If other rules apply along with this one, the one with the highest numeric priority is
+    /// evaluated. If multiple applicable rules have the same priority, that's an overlap error.
+    pub prio: i64,
+    /// If this rule applies, the top-level term should evaluate to this expression.
+    pub result: ExprId,
+}
+
+/// Records whether a given pair of rules can both match on some input.
+pub enum Overlap {
+    /// There is no input on which this pair of rules can both match.
+    No,
+    /// There is at least one input on which this pair of rules can both match.
+    Yes {
+        /// True if every input accepted by one rule is also accepted by the other. This does not
+        /// indicate which rule is more general and in fact the rules could match exactly the same
+        /// set of inputs. You can work out which by comparing the number of constraints in both
+        /// rules: The more general rule has fewer constraints.
+        subset: bool,
+    },
+}
+
+/// A collection of [Rule]s, along with hash-consed [Binding]s and [Expr]s for all of them.
+#[derive(Debug, Default)]
+pub struct RuleSet {
+    /// The [Rule]s for a single [sema::Term].
+    pub rules: Vec<Rule>,
+    /// The bindings identified by [BindingId]s within rules.
+    pub bindings: Vec<Binding>,
+    /// The expressions identified by [ExprId]s within rules.
+    pub exprs: Vec<Expr>,
+}
+
+/// Construct a [RuleSet] for each term in `termenv` that has rules.
+pub fn build(
+    termenv: &sema::TermEnv,
+    tyenv: &sema::TypeEnv,
+) -> (Vec<(sema::TermId, RuleSet)>, Vec<Error>) {
+    let mut errors = Vec::new();
+    let mut term = HashMap::new();
+    for rule in termenv.rules.iter() {
+        term.entry(rule.lhs.root_term().unwrap())
+            .or_insert_with(RuleSetBuilder::default)
+            .add_rule(rule, termenv, tyenv, &mut errors);
+    }
+
+    // The `term` hash map may return terms in any order. Sort them to ensure that we produce the
+    // same output every time when given the same ISLE source. Rules are added to terms in `RuleId`
+    // order, so it's not necessary to sort within a `RuleSet`.
+    let mut result: Vec<_> = term
+        .into_iter()
+        .map(|(term, builder)| (term, builder.rules))
+        .collect();
+    result.sort_unstable_by_key(|(term, _)| *term);
+
+    (result, errors)
+}
+
+impl Rule {
+    /// Returns whether a given pair of rules can both match on some input, and if so, whether
+    /// either matches a subset of the other's inputs.
+    pub fn may_overlap(&self, other: &Rule) -> Overlap {
+        // The outer loop needs to go over the rule with fewer constraints in order to correctly
+        // identify if it's a subset of the other rule. Also, that way around is faster.
```

Comment:
Can you say why this is needed for correctness in this comment, for clarity?

(Because the iteration pattern below catches (present in small, not present in big) and we would miss that case if iterating over `big` in the outermost loop, right?)